pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,458 @@
+type sender struct {
+	cfg  *senderConfig
+	host string
+
+	climit chan struct{} // acts as a semaphore for limiting concurrent connections
+
+	mu        sync.Mutex  // guards below fields
+	list      *list.List  // send queue
+	size      int         // size of send queue
+	scheduled bool        // reports if a flush is scheduled
+	attempt   int         // retry attempt coming up next
+	timer     *time.Timer // timed flush (triggered by backoff)
+}
+
+// senderConfig specifies the configuration for the sender.
+type senderConfig struct {
+	// client specifies the HTTP client to use when sending requests.
+	client *http.Client
+	// url specifies the URL to send requests too.
+	url *url.URL
+	// apiKey specifies the Datadog API key to use.
+	apiKey string
+	// maxConns specifies the maximum number of allowed concurrent ougoing
+	// connections.
+	maxConns int
+	// recorder specifies the eventRecorder to use when reporting events occurring
+	// in the sender.
+	recorder eventRecorder
+}
+
+// newSender creates a new sender, which uses the given HTTP client, URL and API key
+// to make requests using incoming payloads. The limit of outgoing concurrent connections
+// is defined by climit.
+func newSender(cfg *senderConfig) *sender {
+	return &sender{
+		cfg:    cfg,
+		climit: make(chan struct{}, cfg.maxConns),
+		list:   list.New(),
+	}
+}
+
+// Push pushes p onto the sender, to be written to the destination.
+func (q *sender) Push(p *payload) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// enqueue the payload
+	q.enqueueLocked(p)
+
+	if !q.scheduled {
+		// no flush is scheduled; start one
+		q.scheduleFlushLocked(0)
+	}
+}
+
+// scheduleFlushLocked schedules the next flush using the given delay.
+func (q *sender) scheduleFlushLocked(delay time.Duration) {
+	switch {
+	case delay == 0:
+		go q.flush()
+	case q.timer == nil:
+		q.timer = time.AfterFunc(delay, q.flush)
+	default:
+		q.timer.Stop()
+		q.timer.Reset(delay)
+	}
+	q.scheduled = true
+}
+
+// flush drains and sends the entire queue. If anything comes in while flushing
+// or if some of the payloads fail to send as retriable, further follow up flushes
+// are scheduled.
+func (q *sender) flush() {
+	startTime := time.Now()
+
+	// we drain the queue, which is a blocking operation, meaning that
+	// new payloads are stopped from coming in while this happens; but
+	// it is a fast.
+	payloads := q.drainQueue()
+
+	// we send the payloads we've retrieved; while we do this, more payloads
+	// can join the list.
+	done, retries := q.sendPayloads(payloads)
+	if done > 0 {
+		q.recordEvent(eventTypeFlushed, &eventData{
+			count:    int(done),
+			duration: time.Since(startTime),
+		})
+	}
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	switch {
+	case retries > 0:
+		// some sends failed as retriable; we need to back off a bit
+		q.attempt++
+		delay := backoffDuration(q.attempt)
```

Comment:
> possible danger - if the queue is really backed up due to retries, we may end up doing huge bursts of requests here (up to the connection limit). I don't see any kind of rate limit, so we could potentially send out 64 * 1024 * 1024 payloads all at the same time? On smaller hosts that may cause large CPU or network bursts that could negatively impact other things running on the host.

Yep, this is my least favorite aspect this implementation. The rate limit is currently the connection limit, which limits us to maximum 20 (by default) simultaneous connections. I have spent quite some time thinking about a way around this and explored two ideas:

* Do one request first and see if the endpoint has become available again, if yes, do the rest in parallel. I ruled this one out because it added more complexity as well as prolonged the time it takes to flush by 2x or more, only to cover for some very rarely occurring event. Furthermore, if there is only network latency and no real issues sending to the endpoint, it would likely cause the queue to become full and drop payloads due to the slow flushing. Which brings back the original problem we are trying to solve.

* Use a request which [can be canceled](https://golang.org/pkg/context/#WithCancel). This worked, and when one of the payloads failed, it managed to successfully cancel the rest of the parallel requests. However (this took a long time to figure out) I realized that requests can be canceled also after their bodies reach the server, but there was not yet a response, which means that we could sometimes risk sending the same payload twice. And we don't want that.

I'd say it's fine to live with this. We can explore it later if it really becomes a problem. We might be able to reassess the second point from above and perhaps use the [client trace from the `httptrace` package](https://golang.org/pkg/net/http/httptrace/#ClientTrace) to tell if the request body was written to the server by means of the `WroteRequest` field, in cases where the response didn't make it back because of a cancelation.

But for now, let's keep it simple and not account for extreme edge-cases which result in complicated code. The code is already slightly complicated as is. 

> Would a better backoff strategy be simply to have a rate limit per sender? The rate limit defaults to "no limit", then gets gradually bumped up/down as we see errors (or the lack of errors)?

See my answer [here](https://github.com/DataDog/datadog-agent/pull/3688/files#r294979308).

Reply:
> Yep, this is my least favorite aspect of this implementation. The rate limit is currently the connection limit, which limits us to maximum 20 (by default) simultaneous connections.

Oh ok that's not bad. Yeah we're probably fine then.

> * Use a request which [can be canceled](https://golang.org/pkg/context/#WithCancel).

That's probably not necessary. If we've already sent out a request then can just wait for the timeout to cause it to fail. 

I still think a rate-limiter would be a pretty elegant solution that would simplify a lot of this code. I could work something like this

Main loop in each sender just reads payloads in serially of a channel and sends them off asynchronously:
```
case pkg := <- in:
	rateLimiter.acquire()
	sendAsync(pkg)
```

`rateLimiter.acquire()` blocks the right amount of time to enforce a specified rate limit, configured in something like "payloads per second". There's a [nice implementation of this I've used in Google's guava library](https://guava.dev/releases/19.0/api/docs/index.html?com/google/common/util/concurrent/RateLimiter.html).

`sendAsync` will send the payload asynchronously:
* if retryable error, lower the rate limit according to exponential-backoff logic then throw the payload back on the in-channel. Have reasonable minimum for the rate limit, like one payload/second, so it never gets too long. 
* if success, increase the rate limit

If we are fine with dropping payloads when there's more than the configured maximum of pending payloads, we can just use non-blocking channel write to block when the channel is full, and it'll work the same way you're doing now.

Reply:
Right. This is the same to what @AlexJF suggested in another comment. I tried it out and indeed it seems that everything comes out much easier. I will continue to test that and if it’s all good I’ll close this and open a fresh PR.